File: tools/history_belief.py
#!/usr/bin/env python3
"""Offline actor-aware hand inference from a perspective-scrubbed history.

The input may be an omniscient ``analyze``/showcase JSON file, but inference
never receives that object.  ``build_view`` first reduces it to information
the selected observer legitimately has: their initial hand, public actions,
and their own deck draws before the target position.  Opponent deck-draw
identities, opponent hands, future actions, and truth labels are absent from
the wire format accepted by ``bin/history_belief``.

Version 1's transport format is limited to positions at round ply 0..20, but
the actor-specific check is stricter: every preceding action must come from
the deterministic exact-symmetry policy prefix.  The current maintained actor
starts search at ply 14, so its safe targets end at ply 14.  Later rollout
decisions contain private search randomness and require a sequential
likelihood model rather than pretending they are deterministic policy actions.
"""
from __future__ import annotations

import argparse
import hashlib
import json
from pathlib import Path
import subprocess
import sys
from typing import Any


ROOT = Path(__file__).resolve().parents[1]
SUITS = "YBWGR"
DEFAULT_SEED = 0xA7710B311EF


class ViewError(ValueError):
    """The source cannot be reduced to a valid perspective view."""


def parse_card(name: str) -> tuple[int, int]:
    if not isinstance(name, str) or len(name) < 2 or name[0] not in SUITS:
        raise ViewError(f"invalid card name: {name!r}")
    suffix = name[1:]
    if suffix == "x":
        value = 0
    else:
        try:
            value = int(suffix)
        except ValueError as exc:
            raise ViewError(f"invalid card name: {name!r}") from exc
        if not 2 <= value <= 10:
            raise ViewError(f"invalid card value: {name!r}")
    return SUITS.index(name[0]), value


def card_name(suit: int, value: int) -> str:
    return f"{SUITS[suit]}{'x' if value == 0 else value}"


def draw_source(name: str) -> int:
    if name == "deck":
        return 0
    if isinstance(name, str) and len(name) == 1 and name in SUITS:
        return SUITS.index(name) + 1
    raise ViewError(f"invalid draw source: {name!r}")


def canonical_view_bytes(view: dict[str, Any]) -> bytes:
    return json.dumps(
        view, sort_keys=True, separators=(",", ":"), ensure_ascii=True
    ).encode("ascii")


def build_view(
    game: dict[str, Any], target_ply: int, observer: int | None = None
) -> dict[str, Any]:
    """Return the complete, and only, information available to inference."""
    if observer is not None and observer not in (0, 1):
        raise ViewError("observer must be 0 or 1")
    plies = game.get("plies")
    if not isinstance(plies, list):
        raise ViewError("analysis JSON has no ply list")
    try:
        target = next(p for p in plies if int(p["n"]) == target_ply)
    except (StopIteration, KeyError, TypeError, ValueError) as exc:
        raise ViewError(f"target ply {target_ply} is absent") from exc

    round_index = int(target["round"])
    round_ply = int(target["round_ply"])
    if not 0 <= round_ply <= 20:
        raise ViewError(
            "v1 supports only target positions through round ply 20; "
            "later actions may have used stochastic rollout search"
        )
    if observer is None:
        observer = int(target["player"])

    round_records = sorted(
        (p for p in plies if int(p["round"]) == round_index),
        key=lambda p: int(p["round_ply"]),
    )
    if not round_records or int(round_records[0]["round_ply"]) != 0:
        raise ViewError("round does not start at round ply zero")
    if round_ply > len(round_records):
        raise ViewError("round-ply index exceeds recorded round")

    # Reading only hands[observer] is deliberate.  The other hand remains
    # outside the scrubbed view even though the source artifact contains it.
    first = round_records[0]
    hands = first.get("hands")
    if (
        not isinstance(hands, list)
        or len(hands) != 2
        or not isinstance(hands[observer], list)
        or len(hands[observer]) != 8
    ):
        raise ViewError("observer initial hand is malformed")
    own_initial = sorted(
        (parse_card(name) for name in hands[observer]),
        key=lambda item: (item[0], item[1]),
    )

    prefix = [p for p in round_records if int(p["round_ply"]) < round_ply]
    if [int(record["round_ply"]) for record in prefix] != list(range(round_ply)):
        raise ViewError("round history has a gap before target")

    events: list[dict[str, Any]] = []
    expected_actor = int(first["player"])
    for record in prefix:
        actor = int(record["player"])
        if actor != expected_actor:
            raise ViewError("recorded actors do not alternate")
        expected_actor ^= 1
        move = record.get("move")
        if not isinstance(move, dict):
            raise ViewError("recorded move is malformed")
        suit, value = parse_card(move.get("card"))
        action = move.get("act")
        if action not in ("play", "discard"):
            raise ViewError(f"invalid action: {action!r}")
        draw = draw_source(move.get("draw"))
        event: dict[str, Any] = {
            "actor": actor,
            "card": [suit, value],
            "discard": int(action == "discard"),
            "draw": draw,
        }
        # This is the sole private-history field.  Never even access "drawn"
        # for an opponent deck draw.
        if actor == observer and draw == 0:
            event["own_draw"] = list(parse_card(move.get("drawn")))
        events.append(event)

    cum = first.get("cum")
    if (
        not isinstance(cum, list)
        or len(cum) != 2
        or not all(isinstance(x, int) for x in cum)
    ):
        raise ViewError("round cumulative score is malformed")

    return {
        "schema": "lc-perspective-history-v1",
        "observer": observer,
        "round": round_index,
        "start_turn": int(first["player"]),
        "cum": [int(cum[0]), int(cum[1])],
        "target_global_ply": int(target["n"]),
        "target_round_ply": round_ply,
        "own_initial": [list(card) for card in own_initial],
        "events": events,
    }


def worker_wire(view: dict[str, Any]) -> str:
    """Serialize only the scrubbed schema understood by the C worker."""
    lines = ["LCBH1"]
    lines.append(
        f"{view['observer']} {view['round']} {view['start_turn']} "
        f"{view['cum'][0]} {view['cum'][1]} {len(view['events'])}"
    )
    own = view["own_initial"]
    lines.append(
        str(len(own))
        + "".join(f" {int(suit)} {int(value)}" for suit, value in own)
    )
    for event in view["events"]:
        known = event.get("own_draw", [-1, -1])
        lines.append(
            f"{event['actor']} {event['card'][0]} {event['card'][1]} "
            f"{event['discard']} {event['draw']} {known[0]} {known[1]}"
        )
    return "\n".join(lines) + "\n"


def validate_actor_prefix(
    actor_spec: Any,
    target_round_ply: int,
    symmetries: int,
    net_path: Path,
) -> str:
    """Prove that all observed opponent actions used the policy directly."""
    if not isinstance(actor_spec, str):
        raise ViewError("analysis metadata does not identify its actor")
    fields = actor_spec.split(":")
    kind = fields[0]
    if len(fields) < 2 or not fields[1]:
        raise ViewError("source actor spec does not identify its checkpoint")
    source_net_path = Path(fields[1])
    if not source_net_path.is_absolute():
        source_net_path = ROOT / source_net_path
    try:
        source_hash = hashlib.sha256(source_net_path.read_bytes()).digest()
        inference_hash = hashlib.sha256(net_path.read_bytes()).digest()
    except OSError as exc:
        raise ViewError(f"cannot verify source actor checkpoint: {exc}") from exc
    if source_hash != inference_hash:
        raise ViewError(
            "inference checkpoint does not match the source actor checkpoint"
        )

    def numeric(index: int, default: float, label: str) -> float:
        if index >= len(fields):
            return default
        try:
            return float(fields[index])
        except ValueError as exc:
            raise ViewError(
                f"source actor has malformed {label}: {fields[index]!r}"
            ) from exc

    try:
        if kind == "policy":
            temperature = numeric(2, 0.0, "policy temperature")
            actor_symmetries = int(numeric(3, 1.0, "policy symmetries"))
            plan_deck_max = int(numeric(4, 0.0, "planner deck threshold"))
            plan_block_gap = int(numeric(5, 0.0, "planner block gap"))
            if target_round_ply > 0 and temperature != 0.0:
                raise ViewError(
                    "target prefix used stochastic policy sampling; v1 "
                    "requires deterministic argmax actions"
                )
            if (
                target_round_ply > 0
                and plan_deck_max > 0
                and plan_block_gap > 0
            ):
                raise ViewError(
                    "target prefix comes from a planner-enabled policy; v1 "
                    "models only unmodified argmax actions"
                )
        elif kind in ("rollout", "rolloutu"):
            search_from = int(numeric(7, 0.0, "rollout ply threshold"))
            actor_symmetries = int(
                numeric(15, 1.0, "rollout policy symmetries")
            )
            plan_deck_max = int(
                numeric(23, 0.0, "rollout planner deck threshold")
            )
            plan_block_gap = int(
                numeric(24, 0.0, "rollout planner block gap")
            )
            semantic_candidates = int(
                numeric(25, 0.0, "semantic-candidate flag")
            )
            if target_round_ply > search_from:
                raise ViewError(
                    "target prefix includes a rollout-search action; v1 can "
                    "infer only the deterministic policy prefix"
                )
            if (
                target_round_ply > 0
                and plan_deck_max > 0
                and plan_block_gap > 0
            ):
                raise ViewError(
                    "target prefix comes from a planner-enabled rollout actor; "
                    "v1 models only unmodified argmax actions"
                )
            if target_round_ply > 0 and semantic_candidates:
                raise ViewError(
                    "target prefix comes from a semantic-search-enabled "
                    "rollout actor; v1 cannot integrate over that search"
                )
        else:
            raise ViewError(f"unsupported source actor: {kind!r}")
    except ViewError:
        raise
    except (OverflowError, ValueError) as exc:
        raise ViewError("source actor spec has malformed numeric fields") from exc
    if actor_symmetries != symmetries:
        raise ViewError(
            f"source actor used {actor_symmetries} policy symmetries, "
            f"but inference requested {symmetries}"
        )
    return actor_spec


def annotate(
    game: dict[str, Any],
    *,
    target_ply: int,
    observer: int | None,
    worlds: int,
    seed: int,
    symmetries: int,
    net_path: Path,
    worker_path: Path,
) -> dict[str, Any]:
    if not 0 <= seed <= (1 << 64) - 1:
        raise ViewError("annotation seed must fit an unsigned 64-bit integer")
    view = build_view(game, target_ply, observer)
    view_bytes = canonical_view_bytes(view)
    if not worker_path.is_file():
        raise ViewError(
            f"worker not found at {worker_path}; run `make bin/history_belief`"
        )
    if not net_path.is_file():
        raise ViewError(f"network not found at {net_path}")

    meta = game.get("meta") if isinstance(game.get("meta"), dict) else {}
    source_actor = validate_actor_prefix(
        meta.get("actor"), view["target_round_ply"], symmetries, net_path
    )

    command = [
        str(worker_path),
        "-n",
        str(net_path),
        "-w",
        str(worlds),
        "-s",
        str(seed),
        "-y",
        str(symmetries),
    ]
    completed = subprocess.run(
        command,
        input=worker_wire(view),
        text=True,
        capture_output=True,
        cwd=ROOT,
    )
    if completed.returncode != 0:
        detail = completed.stderr.strip() or "worker failed without diagnostics"
        raise ViewError(detail)
    try:
        result = json.loads(completed.stdout)
    except json.JSONDecodeError as exc:
        raise ViewError("worker emitted invalid JSON") from exc

    for card in result.get("cards", []):
        card["card"] = card_name(int(card["suit"]), int(card["value"]))

    result["provenance"] = {
        "information_contract": (
            "observer initial hand + public prefix + observer deck draws; "
            "opponent deck draws, opponent hands, truth labels, and future "
            "actions excluded"
        ),
        "target_global_ply": view["target_global_ply"],
        "target_round_ply": view["target_round_ply"],
        "observer": view["observer"],
        "round": view["round"],
        "source_actor_spec": source_actor,
        "inference_actor": (
            f"policy:{net_path}:0:{symmetries} "
            "(source actor's pre-search behavior)"
        ),
        "model_sha256": hashlib.sha256(net_path.read_bytes()).hexdigest(),
        "view_sha256": hashlib.sha256(view_bytes).hexdigest(),
        "annotation_seed": seed,
        "selection": "all sampled worlds; no posterior screening or retries",
    }
    return result


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("analysis", type=Path)
    parser.add_argument("--ply", required=True, type=int)
    parser.add_argument("--observer", type=int, choices=(0, 1))
    parser.add_argument("--worlds", type=int, default=20000)
    parser.add_argument("--seed", type=int, default=DEFAULT_SEED)
    parser.add_argument("--symmetries", type=int, default=20)
    parser.add_argument("--net", type=Path, default=ROOT / "data/champion.bin")
    parser.add_argument(
        "--worker", type=Path, default=ROOT / "bin/history_belief"
    )
    parser.add_argument(
        "--dump-view",
        action="store_true",
        help="emit only the perspective-scrubbed view; do not run inference",
    )
    parser.add_argument("-o", "--output", type=Path)
    return parser.parse_args()


def main() -> int:
    args = parse_args()
    try:
        game = json.loads(args.analysis.read_text(encoding="utf-8"))
        if args.dump_view:
            output: Any = build_view(game, args.ply, args.observer)
        else:
            output = annotate(
                game,
                target_ply=args.ply,
                observer=args.observer,
                worlds=args.worlds,
                seed=args.seed,
                symmetries=args.symmetries,
                net_path=args.net.resolve(),
                worker_path=args.worker.resolve(),
            )
    except (OSError, json.JSONDecodeError, ViewError) as exc:
        print(f"history_belief.py: {exc}", file=sys.stderr)
        return 1

    rendered = json.dumps(output, indent=2, sort_keys=True) + "\n"
    if args.output:
        temporary = args.output.with_name(f".{args.output.name}.tmp")
        temporary.write_text(rendered, encoding="utf-8")
        temporary.replace(args.output)
    else:
        sys.stdout.write(rendered)
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: tools/history_belief.c
/*
 * history_belief -- actor-aware hidden-hand inference from a scrubbed view.
 *
 * This is deliberately a policy-prefix tool, not a generic omniscient replay.
 * Its stdin contract contains only:
 *   - the observer's initial hand,
 *   - public actions up to the target position, and
 *   - identities of deck cards drawn by the observer.
 *
 * An opponent deck draw has no identity in the input.  Candidate deals are
 * sampled conditional on the observer's legitimate private information, then
 * retained only when the frozen, deterministic symmetry-ensemble policy
 * reproduces every observed opponent action.  The result is a Monte Carlo
 * posterior for the known AI actor, suitable for offline analysis of early
 * positions.  Search-time actions are intentionally outside this contract:
 * integrating over rollout RNG is a different inference problem.
 */
#include "../src/agent.h"
#include "../src/lc.h"
#include "../src/net.h"
#include <errno.h>
#include <math.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define VIEW_MAGIC "LCBH1"
#define MAX_VIEW_EVENTS 20
#define SEM_VALUE_SLOTS 11

typedef struct {
    int actor;
    int suit;
    int value;
    int discard;
    int draw;
    int known_suit;
    int known_value;
    int known_card;
} PublicEvent;

typedef struct {
    int observer;
    int round;
    int start_turn;
    int cum[2];
    int nevent;
    int own_n;
    int own_suit[HAND_SIZE];
    int own_value[HAND_SIZE];
    PublicEvent event[MAX_VIEW_EVENTS];
} View;

static void fail(const char *message)
{
    fprintf(stderr, "history_belief: %s\n", message);
    exit(EXIT_FAILURE);
}

static int valid_semantic(int suit, int value)
{
    return suit >= 0 && suit < NSUIT
        && (value == 0 || (value >= 2 && value <= 10));
}

static int semantic_match(int card, int suit, int value)
{
    if (CARD_SUIT(card) != suit) return 0;
    return value == 0 ? CARD_IS_WAGER(card)
                      : (!CARD_IS_WAGER(card) && CARD_VALUE(card) == value);
}

/* Assign an arbitrary physical ID to an observer-known semantic card.  The
 * maintained champion ties all three wager copies exactly; the input and
 * output remain semantic, so this arbitrary representative is unobservable. */
static int allocate_semantic(int suit, int value, uint64_t used)
{
    if (!valid_semantic(suit, value)) return -1;
    if (value != 0) {
        int card = CARD_MAKE(suit, value + 1);
        return ((used >> card) & 1ULL) ? -1 : card;
    }
    for (int rank = 0; rank < WAGERS_PER_SUIT; rank++) {
        int card = CARD_MAKE(suit, rank);
        if (!((used >> card) & 1ULL)) return card;
    }
    return -1;
}

static void read_view(View *view)
{
    char magic[16];
    memset(view, 0, sizeof *view);
    if (scanf("%15s", magic) != 1 || strcmp(magic, VIEW_MAGIC) != 0)
        fail("stdin is not an LCBH1 perspective view");
    if (scanf("%d%d%d%d%d%d",
              &view->observer, &view->round, &view->start_turn,
              &view->cum[0], &view->cum[1], &view->nevent) != 6)
        fail("short perspective-view header");
    if (view->observer < 0 || view->observer > 1)
        fail("observer must be 0 or 1");
    if (view->round < 0 || view->round >= MATCH_ROUNDS)
        fail("round is outside the supported match");
    if (view->start_turn < 0 || view->start_turn > 1)
        fail("start_turn must be 0 or 1");
    if (view->cum[0] < INT16_MIN || view->cum[0] > INT16_MAX ||
        view->cum[1] < INT16_MIN || view->cum[1] > INT16_MAX)
        fail("cumulative score does not fit State");
    if (view->nevent < 0 || view->nevent > MAX_VIEW_EVENTS)
        fail("only policy-prefix views of at most 20 actions are supported");

    if (scanf("%d", &view->own_n) != 1 || view->own_n != HAND_SIZE)
        fail("own initial hand must contain exactly eight cards");
    for (int i = 0; i < view->own_n; i++) {
        if (scanf("%d%d", &view->own_suit[i], &view->own_value[i]) != 2 ||
            !valid_semantic(view->own_suit[i], view->own_value[i]))
            fail("invalid card in own initial hand");
    }

    int expected_actor = view->start_turn;
    for (int i = 0; i < view->nevent; i++) {
        PublicEvent *e = &view->event[i];
        if (scanf("%d%d%d%d%d%d%d",
                  &e->actor, &e->suit, &e->value, &e->discard, &e->draw,
                  &e->known_suit, &e->known_value) != 7)
            fail("short public event");
        e->known_card = -1;
        if (e->actor != expected_actor)
            fail("event actors do not alternate from start_turn");
        expected_actor ^= 1;
        if (!valid_semantic(e->suit, e->value))
            fail("invalid public action card");
        if ((e->discard != 0 && e->discard != 1) ||
            e->draw < 0 || e->draw > NSUIT)
            fail("invalid public action disposition or draw source");

        if (e->draw == 0 && e->actor == view->observer) {
            if (!valid_semantic(e->known_suit, e->known_value))
                fail("observer deck draw must include its known card");
        } else if (e->known_suit != -1 || e->known_value != -1) {
            fail("opponent deck-draw identities are forbidden");
        }
    }

    char trailing[2];
    if (scanf("%1s", trailing) == 1)
        fail("trailing data after perspective view");
}

static int observed_move(const State *st, const PublicEvent *event, Move *out)
{
    Move moves[MAX_MOVES];
    int n = lc_moves(st, moves);
    for (int i = 0; i < n; i++) {
        if (semantic_match(moves[i].card, event->suit, event->value) &&
            moves[i].discard == event->discard &&
            moves[i].draw == event->draw) {
            *out = moves[i];
            return 1;
        }
    }
    return 0;
}

static Move policy_argmax(const Net *net, const State *st, int symmetries)
{
    Move moves[MAX_MOVES];
    float probability[MAX_MOVES];
    int n = policy_probs_sym(net, st, moves, probability, NULL, symmetries);
    if (n <= 0) fail("policy encountered a state with no legal move");
    int best = 0;
    for (int i = 1; i < n; i++)
        if (probability[i] > probability[best]) best = i;
    return moves[best];
}

static int same_semantic_move(Move move, const PublicEvent *event)
{
    return semantic_match(move.card, event->suit, event->value)
        && move.discard == event->discard
        && move.draw == event->draw;
}

static void shuffle(uint8_t *cards, int n, Rng *rng)
{
    for (int i = n - 1; i > 0; i--) {
        uint32_t j = rng_below(rng, (uint32_t)i + 1);
        uint8_t tmp = cards[i];
        cards[i] = cards[j];
        cards[j] = tmp;
    }
}

static int supported_symmetries(int n)
{
    return n == 1 || n == 5 || n == 10 || n == 20 || n == 120;
}

int main(int argc, char **argv)
{
    const char *net_path = "data/champion.bin";
    long worlds = 20000;
    uint64_t seed = UINT64_C(11506497556975);
    int symmetries = 20;

    for (int i = 1; i < argc; i++) {
        if (!strcmp(argv[i], "-n") && i + 1 < argc) {
            net_path = argv[++i];
        } else if (!strcmp(argv[i], "-w") && i + 1 < argc) {
            char *end = NULL;
            errno = 0;
            worlds = strtol(argv[++i], &end, 10);
            if (errno || !end || *end) fail("invalid world count");
        } else if (!strcmp(argv[i], "-s") && i + 1 < argc) {
            char *end = NULL;
            errno = 0;
            const char *text = argv[++i];
            if (text[0] == '-') fail("invalid annotation seed");
            seed = strtoull(text, &end, 10);
            if (errno || !end || end == text || *end)
                fail("invalid annotation seed");
        } else if (!strcmp(argv[i], "-y") && i + 1 < argc) {
            symmetries = atoi(argv[++i]);
        } else {
            fprintf(stderr,
                    "usage: %s [-n NET] [-w WORLDS] [-s SEED] "
                    "[-y SYMMETRIES] < scrubbed-view\n", argv[0]);
            return EXIT_FAILURE;
        }
    }
    if (worlds < 1 || worlds > 10000000L)
        fail("world count must be between 1 and 10,000,000");
    if (!supported_symmetries(symmetries))
        fail("symmetries must be 1, 5, 10, 20, or 120");

    View view;
    read_view(&view);

    uint64_t private_mask = 0;
    int own_card[HAND_SIZE];
    for (int i = 0; i < HAND_SIZE; i++) {
        own_card[i] = allocate_semantic(view.own_suit[i],
                                        view.own_value[i], private_mask);
        if (own_card[i] < 0)
            fail("own initial hand contains too many copies of a card");
        private_mask |= 1ULL << own_card[i];
    }

    int forced[NCARD];
    for (int i = 0; i < NCARD; i++) forced[i] = -1;
    int deck_draws = 0;
    for (int i = 0; i < view.nevent; i++) {
        PublicEvent *event = &view.event[i];
        if (event->draw != 0) continue;
        int position = 2 * HAND_SIZE + deck_draws++;
        if (position >= NCARD) fail("too many deck draws in view");
        if (event->actor != view.observer) continue;
        int card = allocate_semantic(event->known_suit,
                                     event->known_value, private_mask);
        if (card < 0)
            fail("observer draw repeats a card already in its private record");
        private_mask |= 1ULL << card;
        event->known_card = card;
        forced[position] = card;
    }

    Net *net = malloc(sizeof *net);
    if (!net) fail("out of memory allocating network");
    if (net_load(net, net_path) != 0)
        fail("could not load network");

    double sum[NSUIT][SEM_VALUE_SLOTS] = {{ 0 }};
    double sumsq[NSUIT][SEM_VALUE_SLOTS] = {{ 0 }};
    long accepted = 0;
    Rng rng;
    rng_seed(&rng, seed);

    for (long world = 0; world < worlds; world++) {
        uint8_t pool[NCARD];
        int npool = 0;
        for (int card = 0; card < NCARD; card++)
            if (!((private_mask >> card) & 1ULL))
                pool[npool++] = (uint8_t)card;
        shuffle(pool, npool, &rng);

        uint8_t deck[NCARD];
        memset(deck, 0xff, sizeof deck);
        int cursor = 0;
        int observer_start = view.observer * HAND_SIZE;
        int opponent_start = (view.observer ^ 1) * HAND_SIZE;
        for (int i = 0; i < HAND_SIZE; i++) {
            deck[observer_start + i] = (uint8_t)own_card[i];
            deck[opponent_start + i] = pool[cursor++];
        }
        for (int position = 2 * HAND_SIZE; position < NCARD; position++) {
            if (forced[position] >= 0)
                deck[position] = (uint8_t)forced[position];
            else
                deck[position] = pool[cursor++];
        }
        if (cursor != npool) fail("internal conditional-deal accounting error");

        State state;
        lc_deal_from_deck(&state, deck);
        state.turn = (uint8_t)view.start_turn;
        state.round = (uint8_t)view.round;
        state.cum[0] = (int16_t)view.cum[0];
        state.cum[1] = (int16_t)view.cum[1];

        int keep = 1;
        for (int i = 0; i < view.nevent; i++) {
            PublicEvent *event = &view.event[i];
            if (state.turn != event->actor) {
                keep = 0;
                break;
            }
            Move move;
            if (event->actor == view.observer) {
                if (!observed_move(&state, event, &move)) {
                    keep = 0;
                    break;
                }
            } else {
                move = policy_argmax(net, &state, symmetries);
                if (!same_semantic_move(move, event)) {
                    keep = 0;
                    break;
                }
            }

            if (event->known_card >= 0 &&
                (state.deck_pos >= NCARD ||
                 state.deck[state.deck_pos] != event->known_card)) {
                fail("internal observer-draw conditioning error");
            }
            lc_apply(&state, move);
        }
        if (!keep) continue;

        int opponent = view.observer ^ 1;
        if (__builtin_popcountll(state.hand[opponent]) != HAND_SIZE)
            fail("accepted world has wrong opponent hand size");
        if ((state.hand[opponent] & state.known[opponent])
            != state.known[opponent])
            fail("accepted world dropped a publicly known opponent card");

        accepted++;
        int count[NSUIT][SEM_VALUE_SLOTS] = {{ 0 }};
        uint64_t hand = state.hand[opponent];
        while (hand) {
            int card = __builtin_ctzll(hand);
            hand &= hand - 1;
            int suit = CARD_SUIT(card);
            int value = CARD_IS_WAGER(card) ? 0 : CARD_VALUE(card);
            count[suit][value]++;
        }
        for (int suit = 0; suit < NSUIT; suit++) {
            for (int value = 0; value < SEM_VALUE_SLOTS; value++) {
                if (value == 1) continue;
                double x = count[suit][value];
                sum[suit][value] += x;
                sumsq[suit][value] += x * x;
            }
        }
    }

    free(net);
    if (accepted == 0) fail("no sampled world reproduced the public prefix");

    double marginal_sum = 0.0;
    for (int suit = 0; suit < NSUIT; suit++)
        for (int value = 0; value < SEM_VALUE_SLOTS; value++)
            if (value != 1)
                marginal_sum += sum[suit][value] / (double)accepted;

    printf("{\"schema\":\"lc-history-belief-worker-v1\","
           "\"method\":\"actor_aware_rejection\","
           "\"scope\":\"early_policy_prefix\","
           "\"worlds\":%ld,\"accepted\":%ld,\"accepted_rate\":%.9f,"
           "\"seed\":%llu,\"symmetries\":%d,"
           "\"opponent_hand_size\":%d,\"marginal_sum\":%.9f,"
           "\"cards\":[",
           worlds, accepted, (double)accepted / (double)worlds,
           (unsigned long long)seed, symmetries, HAND_SIZE, marginal_sum);

    int first = 1;
    for (int suit = 0; suit < NSUIT; suit++) {
        for (int value = 0; value < SEM_VALUE_SLOTS; value++) {
            if (value == 1) continue;
            double mean = sum[suit][value] / (double)accepted;
            double variance = 0.0;
            if (accepted > 1) {
                double centered = sumsq[suit][value]
                                - sum[suit][value] * sum[suit][value]
                                  / (double)accepted;
                if (centered < 0.0 && centered > -1e-9) centered = 0.0;
                variance = centered / (double)(accepted - 1);
            }
            double se = sqrt(variance / (double)accepted);
            double maximum = value == 0 ? WAGERS_PER_SUIT : 1.0;
            double lo = mean - 1.96 * se;
            double hi = mean + 1.96 * se;
            if (lo < 0.0) lo = 0.0;
            if (hi > maximum) hi = maximum;
            printf("%s{\"suit\":%d,\"value\":%d,"
                   "\"expected_count\":%.9f,\"se\":%.9f,"
                   "\"ci95\":[%.9f,%.9f]}",
                   first ? "" : ",", suit, value, mean, se, lo, hi);
            first = 0;
        }
    }
    printf("]}\n");
    return EXIT_SUCCESS;
}
